### PR TITLE
Bert

### PR DESCRIPTION
**Pull request by dding3**
Currently only support scala version, will add python version later.

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 1]
```diff
@@ -231,4 +217,36 @@ object TransformerLayer {
     new TransformerLayer[T](nBlock,
       residPdrop, attnPdrop, nHead, maskAttention, embeddingLayer = embeddingLayer)
   }
+
+  def layerNorm[@specialized(Float, Double) T: ClassTag](x: Variable[T],
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: Another `layerNorm` is also defined at line 94.

**Comment by jason-dai** [comment]
Is it possible for the BERT layer to be a subclass of Transformer layer? They only have a little difference:
1) Transformer layer uses `Conv1D` for full connection while BERT layer use `Dense`
2) BERT layer has an additional `attentionMask` argument used in `attn`; we can have the `attn` method check if `attentionMask` is NULL
3) BERT layer is Activity -> Activity, while Transformer layer is Tensor -> Tensor; however, it may make more sense if the input of Transformer layer is a table (input, position), and it can optionally return all block output just like BERT layer.

**Comment by dding3** [comment]
Updated extend `Bert` from `Transformer` for scala version. Will update `Transformer` python version in this PR after scala version is finalized.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+ * @param nBlock block number
+ * @param nHead head number
+ * @param intermediateSize The size of the "intermediate" (i.e., feed-forward)
+ * @param hiddenPDrop The dropout probabilitiy for all fully connected layers
+ * @param attnPDrop drop probability of attention
+ * @param outputAllBlock whether output all blocks' output
+ * @param embeddingLayer embedding layer
+ * @param inputShape input shape, default is null
+ */
+class BERT[T: ClassTag] (
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala]: BERT and Transformer should share the same implementation for doBuild, mlp and multiHeadSelfAttention; they can have different implementations for embedding and the FC layers.

Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala]: Updated.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  nBlock: Int = 12,
+  nHead: Int = 12,
+  intermediateSize: Int = 3072,
+  hiddenPDrop: Double = 0.1,
+  attnPDrop: Double = 0.1,
+  outputAllBlock: Boolean = true,
+  embeddingLayer: KerasLayer[Activity, Tensor[T], T],
+  inputShape: Shape)(implicit ev: TensorNumeric[T])
+  extends TransformerLayer[T](nBlock, hiddenPDrop, attnPDrop, nHead, intermediateSize,
+    true, outputAllBlock, embeddingLayer, inputShape)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala]: mask should be false, no?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 4]
```diff
@@ -37,52 +37,68 @@ import scala.reflect.ClassTag
...
-    seqLen = _inputShape.toSingle().head
+  override def doBuild(inputShape: Shape): AbstractModule[Activity, Activity, T] = {
+    require(inputShape.isInstanceOf[MultiShape], "TransformerLayer input must" +
+      " be a multiple shape")
+    val _inputShape = KerasUtils.removeBatch(inputShape).toMulti()
 
-    val input = Variable(_inputShape)
+    val inputs = _inputShape.map(Variable(_))
+    // requires last input is the attention mask
+    val (extendedAttentionMask, embeddingInputs) = if (inputs.length > 2) {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: we should ask the user to specify if there is a mask?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 5]
```diff
@@ -37,52 +37,68 @@ import scala.reflect.ClassTag
...
+    } else (null, inputs)
+
+    seqLen = _inputShape.head.toSingle().head
 
     require(embeddingLayer.isInstanceOf[Net], "use layers from" +
       "com.intel.analytics.zoo.pipeline.api.keras and operators from" +
       " com.intel.analytics.zoo.pipeline.api.autograd to construct the embedding layer")
     val embedding = embeddingLayer.asInstanceOf[Net]
-    val e = embedding.from(input)
+    val e = embedding.from(embeddingInputs: _*)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: what if there is only one input?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 6]
```diff
@@ -37,52 +37,68 @@ import scala.reflect.ClassTag
...
     with Net {
 
   var seqLen: Int = 0
-  override def doBuild(inputShape: Shape): AbstractModule[Tensor[T], Tensor[T], T] = {
-    require(inputShape.isInstanceOf[SingleShape], "TransformerLayer input must" +
-      " be a single shape")
-    val _inputShape = KerasUtils.removeBatch(inputShape)
-    seqLen = _inputShape.toSingle().head
+  override def doBuild(inputShape: Shape): AbstractModule[Activity, Activity, T] = {
+    require(inputShape.isInstanceOf[MultiShape], "TransformerLayer input must" +
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: TransformerLayer input must be a list of tensors (consisting of input sequence, sequence positions, etc.)

Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: Maybe we should move the checking & input construction to another method, and have BERT layer override the method.
```
(extendedAttentionMask, embeddingInputs) = buildInput(inputShape)
```

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 7]
```diff
@@ -94,61 +110,38 @@ class TransformerLayer[T: ClassTag](
...
-    val u = AutoGrad.mean(x, sizes.size - 1, true)
-    val s = AutoGrad.mean(AutoGrad.square(x - u), sizes.size -1, true)
-    val y = (x - u) / AutoGrad.sqrt(s + e)
-    y * weight + bias
-  }
-
-  def gelu(x: Variable[T]): Variable[T] = {
-    x * 0.5 * (Activation("tanh").from((AutoGrad.square(x) * x * 0.044715 + x)
-      * (scala.math.sqrt(2 / scala.math.Pi))) + 1)
+  def projectLayers(outputSize: Int, config: Double*): Net = {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: projectionLayer

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 8]
```diff
@@ -94,61 +110,38 @@ class TransformerLayer[T: ClassTag](
...
+    new Convolution1D(outputSize, 1, init = RandomNormal(config(0), config(1)))
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: Why not simply `new Convolution1D(outputSize, 1, init = RandomNormal(0.0, 0.02)`?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 9]
```diff
@@ -94,61 +110,38 @@ class TransformerLayer[T: ClassTag](
...
     val key = c.slice(2, hiddenSize, hiddenSize)
     val value = c.slice(2, hiddenSize * 2, hiddenSize)
-    val q = splitHeads(query)
-    val k = splitHeads(key, k = true)
-    val v = splitHeads(value)
-    val a = attn(q, k, v, true) // m: (batch, nhead, seqLen, hiddenSize/nhead)
+    val q = splitHeads(query, nHead)
+    val k = splitHeads(key, nHead, k = true)
+    val v = splitHeads(value, nHead)
+    val a = attn(q, k, v, true, attention_mask) // m: (batch, nhead, seqLen, hiddenSize/nhead)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: a: (batch, nhead, seqLen, hiddenSize/nhead)?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 10]
```diff
@@ -217,18 +242,32 @@ object TransformerLayer {
...
+    outputAllBlock: Boolean,
+    embeddingLayer: KerasLayer[Activity, Tensor[T], T])
     (implicit ev: TensorNumeric[T]): TransformerLayer[T] = {
-    new TransformerLayer[T](nBlock,
-      residPdrop, attnPdrop, nHead, maskAttention, embeddingLayer = embeddingLayer)
+    new TransformerLayer[T](nBlock, residPdrop, attnPdrop, nHead, intermediateSize,
+      maskAttention, outputAllBlock, embeddingLayer = embeddingLayer)
+  }
+
+  def layerNorm[@specialized(Float, Double) T: ClassTag](x: Variable[T],
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: shall we move its definition back to the class?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 11]
```diff
@@ -196,19 +216,24 @@ object TransformerLayer {
     nHead: Int = 12,
     hiddenSize: Int = 768,
     embeddingDrop: Double = 0,
-    maskAttention: Boolean = true)(implicit ev: TensorNumeric[T]): TransformerLayer[T] = {
+    intermediateSize: Int = 3072,
+    maskAttention: Boolean = true,
+    outputAllBlock: Boolean = true)(implicit ev: TensorNumeric[T]): TransformerLayer[T] = {
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: outputAllBlock = false?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 12]
```diff
@@ -94,61 +110,38 @@ class TransformerLayer[T: ClassTag](
...
-  def gelu(x: Variable[T]): Variable[T] = {
-    x * 0.5 * (Activation("tanh").from((AutoGrad.square(x) * x * 0.044715 + x)
-      * (scala.math.sqrt(2 / scala.math.Pi))) + 1)
+  def projectLayers(outputSize: Int, config: Double*): Net = {
+    new Convolution1D(outputSize, 1, init = RandomNormal(config(0), config(1)))
   }
 
   def mlp(x: Variable[T], hiddenSize: Int): Variable[T] = {
-    val h = new Convolution1D(hiddenSize * 4, 1, init = RandomNormal(0.0, 0.02)).from(x)
+    val h = projectLayers(intermediateSize, Array(0.0, 0.02): _*).from(x)
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: why change from `hiddenSize * 4` to `intermediateSize`?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala (hunk 3 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala]: attention_mask which used to mask the padding tokens is also kind of mask?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 6 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: Updated

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 8 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: Updated

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 10 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: layerNorm is also used in when create default embedding layer in Bert.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 12 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: `intermediateSize` is configurable in BERT, to share the same `mlp` with `transformer`  we need change `hiddenSize*4` to `intermediateSize`. Have set `intermediateSize`=`hiddenSize*4` when create `transformer` so existing user will not be impacted by this change.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+ * @param embeddingLayer embedding layer
+ * @param inputShape input shape, default is null
+ */
+class BERT[T: ClassTag] (
+  nBlock: Int = 12,
+  nHead: Int = 12,
+  intermediateSize: Int = 3072,
+  hiddenPDrop: Double = 0.1,
+  attnPDrop: Double = 0.1,
+  initializerRange: Double = 0.02,
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala]: Why do you need to make this configurable? It is not configurable in the previous implementation anyway.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 12 above)
Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: It is not handled if the user does not use default embedding.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala (hunk 3 above)
Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala]: This is confusing. I think you should rename `maskAttention` to `bidirectional`, which indicates if the Transformer layer if bidirectional to unidirectional.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 7 above)
Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: projectionLayer, not projectLayer

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 14]
```diff
@@ -217,18 +246,34 @@ object TransformerLayer {
...
   def apply[@specialized(Float, Double) T: ClassTag](
     nBlock: Int,
     residPdrop: Double,
     attnPdrop: Double,
     nHead: Int,
+    intermediateSize: Int,
+    initializerRange: Double,
     maskAttention: Boolean,
-    embeddingLayer: KerasLayer[Tensor[T], Tensor[T], T])
+    outputAllBlock: Boolean,
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: default to false?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 15]
```diff
@@ -196,19 +219,25 @@ object TransformerLayer {
     nHead: Int = 12,
     hiddenSize: Int = 768,
     embeddingDrop: Double = 0,
-    maskAttention: Boolean = true)(implicit ev: TensorNumeric[T]): TransformerLayer[T] = {
+    initializerRange: Double = 0.02,
+    maskAttention: Boolean = true,
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: why maskAttention is true by default?

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 11 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: Updated

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 12 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: Updated

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala (hunk 13 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/BERT.scala]: I found it's configurable in bert official code and it's also configured in official pre-train model.

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 14 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: Yes, default value is `false`

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 15 above)
Comment (dding3) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: Updated

**Comment by dding3** [comment]
jekins passed. http://10.239.47.211:18888/job/ZOO-PR-Validation/2117/

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala (hunk 14 above)
Comment (jason-dai) [reply zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: it is not default to false in this apply method

File: zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala [hunk 16]
```diff
@@ -125,15 +125,16 @@ class TransformerLayer[T: ClassTag](
   }
 
   def mlp(x: Variable[T], hiddenSize: Int): Variable[T] = {
-    val h = projectLayer(intermediateSize).from(x)
+    val size = if (intermediateSize != 0) intermediateSize else hiddenSize * 4
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/pipeline/api/keras/layers/TransformerLayer.scala]: intermediateSize >0 ?

**Comment by dding3** [comment]
Thanks for your comments.

Jekins passed. http://172.168.2.101:8080/view/ZOO-PR/job/ZOO-PR-Validation/2124/

**Pull request merged**